pkg/policy/repository.go

```diff
@@ -789,27 +757,123 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
+	// enabled for the endpoint.
+	// If the endpoint has the reserved:init label, i.e. if it has not yet
+	// received any labels, always enforce policy (default deny).
+	if policyMode == option.AlwaysEnforce || lbls.Has(labels.IDNameInit) {
 		return true, true, matchingRules
-	case option.DefaultEnforcement:
-		ingress, egress, matchingRules = p.getMatchingRules(securityIdentity)
-		// If the endpoint has the reserved:init label, i.e. if it has not yet
-		// received any labels, always enforce policy (default deny).
-		if lbls.Has(labels.IDNameInit) {
-			return true, true, matchingRules
+	}
+
+	// Determine the default policy for each direction.
+	//
+	// By default, endpoints have no policy and all traffic is allowed.
+	// If any rules select the endpoint, then policy is enabled and non-allowed traffic
+	// is denied.
+	//
+	// Rules, however, can optionally be configured in "pass" mode, where they make
+	// no indication as to the default policy. If an endpoint only has "pass" rules,
+	// then there is no implicit default-deny rule and all traffic **that is not explicitly
+	// denied** is allowed.
+	//
+	// There are three possible cases _per direction_:
+	// 1: No rules are present, or only Allow + Pass rules. Then, policy is disabled.
+	// 2: A non-Pass rule is present. Then, policy is enabled and traffic is default-deny
+	// 3: Only Deny + Pass rules are present. Then, policy is enabled, but we must insert
+	//    an additional allow-all rule. Since wildcard rules have lowest precedence
+	//    in the datapath policy engine, this is a default-allow with specific deny
+	hasIngressDenyPass := false
+	hasEgressDenyPass := false
+	for _, r := range matchingRules {
+		if r.DefaultAction == api.PolicyActionPass {
+			// We need to account for the specific case of a Deny + Pass rule
+			if !hasIngressDenyPass && len(r.IngressDeny) > 0 {
+				hasIngressDenyPass = true
+			}
+
+			if !hasEgressDenyPass && len(r.EgressDeny) > 0 {
+				hasEgressDenyPass = true
+			}
+
+			// Note: we don't care about Allow + Pass rules; they cannot change
+			// the state of the endpoint.
+		} else {
+			// This is a default-Deny rule, so we must enable policy accordingly
+			if !ingress && (len(r.Ingress) > 0 || len(r.IngressDeny) > 0) {
+				ingress = true
+			}
+			if !egress && (len(r.Egress) > 0 || len(r.EgressDeny) > 0) {
+				egress = true
+			}
 		}
+		if ingress && egress {
+			break
+		}
+	}
 
-		// Default mode means that if rules contain labels that match this
-		// endpoint, then enable policy enforcement for this endpoint.
-		return ingress, egress, matchingRules
-	default:
-		// If policy enforcement isn't enabled, we do not enable policy
-		// enforcement for the endpoint. We don't care about returning any
-		// rules that match.
-		return false, false, nil
+	// If there are no default-Deny ingress rules, but there is a pass + deny ingress rule, we must
+	// add an additional wildcard rule for each direction
+	if !ingress && hasIngressDenyPass {
```

Comment:
One awkwardness here is that we conflate "policy is enabled" with "there is a default allow rule". In the case where we want to Deny-with-default-allow, we could also "disable policy" but still return deny rules.

For the BPF policy engine, these are identical. However, Envoy consumes these fields as well. So, in the case where we have only deny+pass, I synthesize an explicit allow rule and set policy to enabled rather than "disable policy".

Reply:
Note: this comment is no longer accurate; L3L4 rules may have L7 policies attached, so we always synthesize a default allow.